File: src/main/java/com/munzenberger/feed/util/XMLFilterReader.java
/*
 * Copyright 2017 Brian Munzenberger
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 * 		http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.munzenberger.feed.util;

import java.io.FilterReader;
import java.io.IOException;
import java.io.Reader;

public class XMLFilterReader extends FilterReader {
	
	private final String encoding;
	private boolean first = true;
	
	public XMLFilterReader(Reader in, String encoding) {
		super(in);
		this.encoding = encoding;
	}

	@Override
	public int read() throws IOException {
		throw new UnsupportedOperationException();
	}

	@Override
	public int read(char[] cbuf, int off, int len) throws IOException {
		int read = super.read(cbuf, off, len);
		
		if (read < 0) {
			return read;
		}
		
		int ptr = off;
		int ctr = 0;
		
		for (int i = off; i < off+read; i++) {
			
			if (first) {
				first = false;
				// strip the UTF-8 byte order mark, if present
				if ("UTF-8".equals(encoding) && cbuf[i] == '\uFEFF') {
					continue;
				}
			}
			
			if (isValidXMLChar(cbuf[i])) {
				cbuf[ptr++] = cbuf[i];
				ctr++;
			}
		}
		
		return ctr;
	}
	
	private static boolean isValidXMLChar(char c) {

		// from https://www.w3.org/TR/xml/#charsets
		return c == 0x9 || c == 0xA || c == 0xD ||
				0x20 <= c && c <= 0xD7FF ||
				0xE000 <= c && c <= 0xFFFD ||
				0x10000 <= c && c <= 0x10FFFF;
	}
}

File: src/main/kotlin/com/munzenberger/feed/source/XMLFilterReader.kt
package com.munzenberger.feed.source

import java.io.FilterReader
import java.io.InputStream
import java.io.InputStreamReader
import java.io.Reader
import java.lang.UnsupportedOperationException

class XMLFilterReader(
        inStream: InputStream,
        private val encoding: String
) : FilterReader(InputStreamReader(inStream, encoding)) {

    companion object {
        private const val BYTE_ORDER_MARK = '\uFEFF'
    }

    private var firstChar: Boolean = true
    private var contentStarted: Boolean = false

    override fun read(): Int {
        throw UnsupportedOperationException()
    }

    override fun reset() {
        throw UnsupportedOperationException()
    }

    override fun markSupported(): Boolean {
        return false
    }

    override fun mark(readAheadLimit: Int) {
        throw UnsupportedOperationException()
    }

    override fun read(cbuf: CharArray, off: Int, len: Int): Int {
        val read = super.read(cbuf, off, len)

        if (read < 0) {
            return read
        }

        var pointer = off
        var counter = 0

        for (i in off until off + read) {

            val c = cbuf[i]

            if (firstChar) {
                firstChar = false
                // strip the UTF-8 byte order mark, if present
                if (encoding == "UTF-8" && c == BYTE_ORDER_MARK) {
                    continue
                }
            }

            if (!contentStarted) {
                // trim any leading whitespace
                if (c.isWhitespace()) {
                    continue
                }
                contentStarted = true
            }

            if (c.isValidXML()) {
                cbuf[pointer++] = c
                counter++
            }
        }

        return counter
    }
}

private fun Char.isValidXML(): Boolean = toInt().let {
    // https://www.w3.org/TR/xml/#charsets
    it == 0x9 ||
            it == 0xA ||
            it == 0xD ||
            it in 0x20..0xD7FF ||
            it in 0xE000..0xFFFD ||
            it in 0x10000..0x10FFFF
}
